propolis/src/instance.rs

```diff
@@ -121,181 +168,283 @@ impl Instance {
-
-        if !State::valid_target(&state.current, &target) {
-            return Err(io::Error::new(
-                io::ErrorKind::InvalidInput,
-                "invalid target",
-            ));
+    pub fn set_target_state(&self, target: ReqState) -> Result<(), ()> {
+        let mut inner = self.inner.lock().unwrap();
+
+        if matches!(inner.state_target, Some(State::Halt | State::Destroy)) {
+            // Cannot request any state once the target is halt/destroy
+            return Err(());
+        }
+        if inner.state_target == Some(State::Reset) && target == ReqState::Run {
+            // Requesting a run when already on the road to reboot is an
+            // immediate success.
+            return Ok(());
+        }
+
+        match target {
+            ReqState::Run => {
+                self.set_target_state_locked(&mut inner, State::Run)
+            }
+            ReqState::Reset => self.trigger_suspend_locked(
+                &mut inner,
+                SuspendKind::Reset,
+                SuspendSource::External,
+            ),
+            ReqState::Halt => self.trigger_suspend_locked(
+                &mut inner,
+                SuspendKind::Halt,
+                SuspendSource::External,
+            ),
+        }
+    }
+
+    pub(crate) fn trigger_suspend(
+        &self,
+        kind: SuspendKind,
+        source: SuspendSource,
+    ) -> Result<(), ()> {
+        let mut inner = self.inner.lock().unwrap();
+        self.trigger_suspend_locked(&mut inner, kind, source)
+    }
+
+    fn trigger_suspend_locked(
+        &self,
+        inner: &mut MutexGuard<Inner>,
+        kind: SuspendKind,
+        source: SuspendSource,
+    ) -> Result<(), ()> {
+        if matches!(inner.state_current, State::Halt | State::Destroy) {
+            // No way out from Halt or Destroy
+            return Err(());
+        }
+
+        match kind {
+            SuspendKind::Reset => {
+                match inner.suspend_info {
+                    Some((SuspendKind::Halt, _)) => {
+                        // Cannot supersede active halt
+                        return Err(());
+                    }
+                    Some((SuspendKind::Reset, _)) => {
+                        return Ok(());
+                    }
+                    _ => {}
+                }
+
+                let hdl = inner.machine.get_hdl();
+                let _ =
+                    hdl.suspend(bhyve_api::vm_suspend_how::VM_SUSPEND_RESET);
+                inner.suspend_info = Some((kind, source));
+                self.set_target_state_locked(inner, State::Reset)
+            }
+            SuspendKind::Halt => {
+                if matches!(inner.suspend_info, Some((SuspendKind::Halt, _))) {
+                    return Ok(());
+                }
+                if inner.suspend_info.is_none() {
+                    let hdl = inner.machine.get_hdl();
+                    let _ = hdl.suspend(
```

Comment:
Why do we only issue the suspend here if there isn't any `suspend_info`?

Reply:
Since these represent hard resets or poweroffs, we don't allow competing suspend requests to be issued.  Once an instance is progressing through a suspend, the bhyve kernel component will reject further requests until the instance undergoes a `VM_REINIT` operation.